scripts/imgtool/image.py

```diff
@@ -0,0 +1,164 @@
+"""
+Image signing and management.
+"""
+
+from . import version as versmod
+import hashlib
+import struct
+
+IMAGE_MAGIC = 0x96f3b83c
+IMAGE_HEADER_SIZE = 32
+
+# Image header flags.
+IMAGE_F = {
+        'PIC':                   0x0000001,
+        'SHA256':                0x0000002,
+        'PKCS15_RSA2048_SHA256': 0x0000004,
+        'ECDSA224_SHA256':       0x0000008,
+        'NON_BOOTABLE':          0x0000010,
+        'ECDSA256_SHA256':       0x0000020,
+        'PKCS1_PSS_RSA2048_SHA256': 0x0000040, }
+
+TLV_VALUES = {
+        'SHA256': 1,
+        'RSA2048': 2,
+        'ECDSA224': 3,
+        'ECDSA256': 4, }
+
+TLV_HEADER_SIZE = 4
+
+# Sizes of the image trailer, depending on image alignment.
+trailer_sizes = {
+        1: 402,
+        2: 788,
+        4: 1560,
+        8: 3104, }
+
+boot_magic = bytes([
+    0x77, 0xc2, 0x95, 0xf3,
+    0x60, 0xd2, 0xef, 0x7f,
+    0x35, 0x52, 0x50, 0x0f,
+    0x2c, 0xb6, 0x79, 0x80, ])
+
+class TLV():
+    def __init__(self):
+        self.buf = bytearray()
+
+    def add(self, kind, payload):
+        """Add a TLV record.  Kind should be a string found in TLV_VALUES above."""
+        buf = struct.pack('<BBH', TLV_VALUES[kind], 0, len(payload))
+        self.buf += buf
+        self.buf += payload
+
+    def get(self):
+        return bytes(self.buf)
+
+class Image():
+    @classmethod
+    def load(cls, path, **kwargs):
+        """Load an image from a given file"""
+        with open(path, 'rb') as f:
+            payload = f.read()
+        obj = cls(**kwargs)
+        obj.payload = payload
+        obj.check()
+        return obj
+
+    def __init__(self, version=None, header_size=IMAGE_HEADER_SIZE, pad=0):
+        self.version = version or versmod.decode_version("0")
+        self.header_size = header_size or IMAGE_HEADER_SIZE
```

Comment:
do_sign should initialize the header_size kwarg appropriately -- putting that into the library is an abstraction violation.